feat(useCanvasContext): new function

michael-observerly opened the pull request:
### Description

So this PR solves the usual setup of a canvas object, whilst simultaneously making it reactive to it's background element (utilising functionality from the useElementBoundin() composable).

Closes #1346 

### Additional context

If this PR is acceptable, I will also extend to another 3D rendering context composable (webgl), as well as PR'ing an integration for three.js if there is an appetite for it.

### What is the purpose of this pull request?

- [ ] Bug fix
- [x] New Feature
- [x] Documentation update
- [ ] Other

### Before submitting the PR, please make sure you do the following

- [x] Read the [Contributing Guidelines](https://github.com/vueuse/vueuse/blob/main/CONTRIBUTING.md).
- [x] Read the [Pull Request Guidelines](https://github.com/vueuse/vueuse/blob/main/packages/guidelines.md).
- [x] Check that there isn't already a PR that solves the problem the same way to avoid creating a duplicate.
- [x] Provide a description in this PR that addresses **what** the PR is solving, or reference the issue that it solves (e.g. `fixes #123`).
- [x] Ideally, include relevant tests that fail without this PR but pass with it.

okxiaoliang4 commented on packages/core/useCanvas2D/demo.vue:
```diff
@@ -0,0 +1,52 @@
+<script setup lang="ts">
+import { computed, reactive, ref } from 'vue-demi'
+import YAML from 'js-yaml'
```
```suggestion
import { stringify } from '@vueuse/docs-utils'
```

okxiaoliang4 commented on packages/core/useCanvas2D/demo.vue:
```diff
@@ -0,0 +1,52 @@
+  height,
+} = useCanvas2D(canvas, bound, { alpha: false })
+
+const properties = computed(() => YAML.dump(
+  reactive({
+    width,
+    height,
+  }),
+))
```
```suggestion
const properties = stringify(reactive({
  width,
  height,
}))
```

okxiaoliang4 commented on packages/core/useCanvas2D/index.ts:
```diff
@@ -0,0 +1,103 @@
+ * Reactive 2D Canvas API
+ *
+ * @param element (HTML Canavs Element) template ref
+ * @param attributes
+ * @param options
```
```suggestion
 * @param bound
 * @param attributes
```

okxiaoliang4 commented on packages/core/useCanvas2D/index.ts:
```diff
@@ -0,0 +1,103 @@
+ */
+export function useCanvas2D(
+  el: Ref<HTMLCanvasElement | null>,
+  bound: Ref<HTMLElement | null>,
```
why we need this?

okxiaoliang4 commented on packages/core/useCanvas2D/index.ts:
```diff
@@ -0,0 +1,103 @@
+    return !!ctx.value
+  })
+
+  const { width: rectWidth, height: rectHeight } = useElementBounding(bound)
+
+  watch(rectWidth, (widthValue) => {
+    if (!canvas.value) return
+    canvas.value.width = widthValue
+  })
+
+  watch(rectHeight, (heightValue) => {
+    if (!canvas.value) return
+    canvas.value.height = heightValue
+  })
```
I think it might be unnecessary, we can use it like this.
```typescript
const canvas = ref<HTMLCanvasElement>()
const { ctx } = useCanvas2D(canvas)

const { width, height } = useElementBounding(canvas)
```

okxiaoliang4 commented on packages/core/useCanvas2D/index.ts:
```diff
@@ -0,0 +1,93 @@
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext
+ */
+export function useCanvas2D(
+  el: Ref<HTMLCanvasElement | null>,
```
```suggestion
  el: MaybeRef<HTMLCanvasElement | null>,
```

okxiaoliang4 commented on packages/core/useCanvas2D/index.ts:
```diff
@@ -0,0 +1,93 @@
+  // Watching the canvas to ensure we can get the canvas' context:
+  watch(canvas, (canvasValue) => {
+    if (!canvasValue) return
+    ctx.value = getContext()
```
```suggest
  attributes?: MaybeRef<UseCanvasAttributes>,
```

michael-observerly commented:
Hey @okxiaoliang4 I've now refactored the code with your suggestions and I have also added dual support for '2d' and 'webgl' contexts as these are the most widely used.

okxiaoliang4 commented:
@michael-observerly Hi, I have initiated a pr in your warehouse to see if this can solve your needs. It has better type hints and less code.
https://github.com/michael-observerly/vueuse/pull/2
